File: scripts/create-local-backup.mjs
#!/usr/bin/env node

import { createReadStream, createWriteStream } from "node:fs";
import { opendir, lstat, mkdir, realpath, readFile, stat } from "node:fs/promises";
import { once } from "node:events";
import { basename, relative, resolve } from "node:path";
import { finished } from "node:stream/promises";
import { createGzip } from "node:zlib";
import { writeSafetyReport } from "./local-safety-report.mjs";

const repoRoot = await realpath(resolve(process.cwd()));
const backupDir = resolve(repoRoot, "local_backups");
const archiveName = `sprintos-local-backup-${timestamp()}.tar.gz`;
const archivePath = resolve(backupDir, archiveName);
const maxScanBytes = 5 * 1024 * 1024;

const blockedDirReasons = new Map([
  [".cache", "caches"],
  [".git", "gitInternals"],
  [".hg", "gitInternals"],
  [".mypy_cache", "caches"],
  [".next", "buildArtifacts"],
  [".nuxt", "buildArtifacts"],
  [".parcel-cache", "caches"],
  [".playwright-cli", "screenshots"],
  [".pytest_cache", "caches"],
  [".ruff_cache", "caches"],
  [".svn", "gitInternals"],
  [".turbo", "caches"],
  [".venv", "dependencies"],
  ["__pycache__", "caches"],
  ["backups", "backupRecursion"],
  ["build", "buildArtifacts"],
  ["coverage", "buildArtifacts"],
  ["data", "localState"],
  ["dist", "buildArtifacts"],
  ["env", "dependencies"],
  ["exports", "runtimeExports"],
  ["htmlcov", "buildArtifacts"],
  ["local_backups", "backupRecursion"],
  ["node_modules", "dependencies"],
  ["out", "buildArtifacts"],
  ["playwright-report", "screenshots"],
  ["screenshots", "screenshots"],
  ["test-results", "screenshots"],
  ["tmp", "caches"],
  ["venv", "dependencies"],
  ["workspaces", "runtimeExports"],
]);

const blockedFileNames = new Set([".DS_Store"]);
const blockedFileSuffixes = [
  ".log",
  ".pyc",
  ".pyo",
  ".sqlite",
  ".sqlite-journal",
  ".sqlite-shm",
  ".sqlite-wal",
  ".tar",
  ".tar.gz",
  ".tgz",
  ".zip",
];

const payloadLogPatterns = [
  /provider[-_ ]?payload/i,
  /provider[-_ ]?response/i,
  /provider[-_ ]?request/i,
  /provider[-_ ]?prompt/i,
  /raw[-_ ]?provider/i,
  /raw[-_ ]?prompt/i,
  /raw[-_ ]?response/i,
  /prompt[-_ ]?payload/i,
  /prompt[-_ ]?log/i,
  /payload[-_ ]?log/i,
  /ai[-_ ]?payload/i,
  /llm[-_ ]?payload/i,
];

const secretPatterns = [
  /\bsk-(?:proj-)?[A-Za-z0-9_-]{20,}\b/,
  /\bDEEPSEEK_API_KEY\s*=\s*["']?[A-Za-z0-9_-]{20,}/i,
  /\bAuthorization\s*:\s*Bearer\s+[A-Za-z0-9._-]{20,}/i,
  /\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{20,}\b/,
  /\bgithub_pat_[A-Za-z0-9_]{30,}\b/,
  /\bAKIA[0-9A-Z]{16}\b/,
  /\bxox[baprs]-[A-Za-z0-9-]{20,}\b/,
];

const safePlaceholderMarkers = ["example", "fake", "local-only", "placeholder", "test", "your_"];
const safePlaceholderPrefixes = [
  "sk-deepseek-",
  "sk-live-",
  "sk-openai-",
  "sk-prototype-",
  "sk-release-",
  "sk-route-",
];

const omitted = {
  envFiles: 0,
  gitInternals: 0,
  dependencies: 0,
  buildArtifacts: 0,
  caches: 0,
  screenshots: 0,
  runtimeExports: 0,
  localState: 0,
  backupRecursion: 0,
  payloadLogs: 0,
  secretMatches: 0,
  localNoise: 0,
  symlinks: 0,
};

await mkdir(backupDir, { recursive: true });

const entries = [];
await collectEntries(repoRoot, "", entries);

if (entries.length === 0) {
  throw new Error("No files were eligible for backup.");
}

await writeTarGz(entries, archivePath);
verifyArchiveEntryList(entries);

const archiveStats = await stat(archivePath);
const includedFiles = entries.filter((entry) => entry.kind === "file").length;

await writeSafetyReport({
  backup: {
    status: "passed",
    archivePath: relative(repoRoot, archivePath),
    archiveSizeBytes: archiveStats.size,
    includedFiles,
    omitted,
  },
});

console.log(`Backup created: ${relative(repoRoot, archivePath)}`);
console.log(`Archive size: ${formatBytes(archiveStats.size)}`);
console.log(`Included files: ${includedFiles}`);
console.log("Omitted files/folders:");
console.log(`- env files: ${omitted.envFiles}`);
console.log(`- git internals: ${omitted.gitInternals}`);
console.log(`- dependencies: ${omitted.dependencies}`);
console.log(`- build artifacts: ${omitted.buildArtifacts}`);
console.log(`- caches: ${omitted.caches}`);
console.log(`- screenshots: ${omitted.screenshots}`);
console.log(`- runtime exports/workspaces: ${omitted.runtimeExports}`);
console.log(`- local SQLite/runtime state: ${omitted.localState}`);
console.log(`- backup recursion: ${omitted.backupRecursion}`);
console.log(`- raw prompts/provider payloads: ${omitted.payloadLogs}`);
console.log(`- files with obvious secret patterns: ${omitted.secretMatches}`);
console.log(`- local noise/symlinks: ${omitted.localNoise + omitted.symlinks}`);
console.log("Verification passed: archive entries contain no blocked env files, git internals, dependencies, build outputs, caches, screenshots, generated payload logs, runtime state, local backups, or obvious secret paths.");

async function collectEntries(absDir, relDir, entries) {
  let dir;
  try {
    dir = await opendir(absDir);
  } catch {
    return;
  }

  for await (const dirent of dir) {
    const relPath = relDir ? `${relDir}/${dirent.name}` : dirent.name;
    const absPath = resolve(repoRoot, relPath);

    if (dirent.isSymbolicLink()) {
      omitted.symlinks += 1;
      continue;
    }

    const exclusion = exclusionFor(relPath, dirent.name);
    if (exclusion) {
      omitted[exclusion] += 1;
      continue;
    }

    if (dirent.isDirectory()) {
      entries.push({ kind: "directory", absPath, relPath, stats: await lstat(absPath) });
      await collectEntries(absPath, relPath, entries);
      continue;
    }

    if (dirent.isFile()) {
      if (await fileContainsSecret(absPath)) {
        omitted.secretMatches += 1;
        continue;
      }
      entries.push({ kind: "file", absPath, relPath, stats: await lstat(absPath) });
      continue;
    }

    omitted.localNoise += 1;
  }
}

function exclusionFor(relPath, name) {
  if (isEnvFile(name)) {
    return "envFiles";
  }
  if (isPayloadLogName(relPath)) {
    return "payloadLogs";
  }
  if (blockedFileNames.has(name) || blockedFileSuffixes.some((suffix) => name.endsWith(suffix))) {
    return "localNoise";
  }
  for (const part of relPath.split("/")) {
    if (blockedDirReasons.has(part)) {
      return blockedDirReasons.get(part);
    }
  }
  return null;
}

function isEnvFile(name) {
  return name === ".env" || name.startsWith(".env.");
}

function isPayloadLogName(name) {
  return payloadLogPatterns.some((pattern) => pattern.test(name));
}

async function fileContainsSecret(absPath) {
  let stats;
  try {
    stats = await lstat(absPath);
  } catch {
    return false;
  }
  if (!stats.isFile() || stats.size === 0 || stats.size > maxScanBytes) {
    return false;
  }
  let buffer;
  try {
    buffer = await readFile(absPath);
  } catch {
    return false;
  }
  if (buffer.subarray(0, Math.min(buffer.length, 8000)).includes(0)) {
    return false;
  }
  const text = buffer.toString("utf8");
  return secretPatterns.some((pattern) => hasNonPlaceholderSecret(pattern, text));
}

function hasNonPlaceholderSecret(pattern, text) {
  const flags = pattern.flags.includes("g") ? pattern.flags : `${pattern.flags}g`;
  const matches = text.match(new RegExp(pattern.source, flags));
  if (!matches) {
    return false;
  }
  return matches.some((match) => !isSafePlaceholderMatch(match));
}

function isSafePlaceholderMatch(value) {
  const normalized = value.toLowerCase();
  if (safePlaceholderMarkers.some((marker) => normalized.includes(marker))) {
    return true;
  }
  const tokenMatch = normalized.match(/sk-[a-z0-9_-]+/);
  return Boolean(tokenMatch && safePlaceholderPrefixes.some((prefix) => tokenMatch[0].startsWith(prefix)));
}

async function writeTarGz(entries, destination) {
  const output = createWriteStream(destination, { mode: 0o600 });
  const gzip = createGzip({ level: 9 });
  gzip.pipe(output);

  for (const entry of entries) {
    await writeGzip(gzip, tarHeader(entry));
    if (entry.kind === "file") {
      await writeFileToGzip(gzip, entry.absPath);
      const padding = paddingFor(entry.stats.size);
      if (padding > 0) {
        await writeGzip(gzip, Buffer.alloc(padding));
      }
    }
  }

  await writeGzip(gzip, Buffer.alloc(1024));
  gzip.end();
  await finished(output);
}

async function writeFileToGzip(gzip, filePath) {
  const input = createReadStream(filePath);
  for await (const chunk of input) {
    await writeGzip(gzip, chunk);
  }
}

async function writeGzip(gzip, chunk) {
  if (!gzip.write(chunk)) {
    await once(gzip, "drain");
  }
}

function tarHeader(entry) {
  const header = Buffer.alloc(512, 0);
  const normalizedName = entry.kind === "directory" && !entry.relPath.endsWith("/")
    ? `${entry.relPath}/`
    : entry.relPath;
  const { name, prefix } = splitTarName(normalizedName);
  const mode = entry.kind === "directory" ? 0o755 : (entry.stats.mode & 0o777) || 0o644;
  const size = entry.kind === "file" ? entry.stats.size : 0;
  const mtime = Math.floor(entry.stats.mtimeMs / 1000);

  writeString(header, name, 0, 100);
  writeOctal(header, mode, 100, 8);
  writeOctal(header, 0, 108, 8);
  writeOctal(header, 0, 116, 8);
  writeOctal(header, size, 124, 12);
  writeOctal(header, mtime, 136, 12);
  header.fill(0x20, 148, 156);
  header[156] = entry.kind === "directory" ? "5".charCodeAt(0) : "0".charCodeAt(0);
  writeString(header, "ustar", 257, 6);
  writeString(header, "00", 263, 2);
  writeString(header, "sprintos", 265, 32);
  writeString(header, "sprintos", 297, 32);
  writeString(header, prefix, 345, 155);

  const checksum = header.reduce((sum, byte) => sum + byte, 0);
  writeOctal(header, checksum, 148, 8);
  return header;
}

function splitTarName(pathName) {
  if (Buffer.byteLength(pathName) <= 100) {
    return { name: pathName, prefix: "" };
  }
  const parts = pathName.split("/");
  for (let index = 1; index < parts.length; index += 1) {
    const prefix = parts.slice(0, index).join("/");
    const name = parts.slice(index).join("/");
    if (Buffer.byteLength(prefix) <= 155 && Buffer.byteLength(name) <= 100) {
      return { name, prefix };
    }
  }
  throw new Error(`Path is too long for portable tar format: ${pathName}`);
}

function writeString(buffer, value, offset, length) {
  buffer.write(value, offset, Math.min(Buffer.byteLength(value), length), "utf8");
}

function writeOctal(buffer, value, offset, length) {
  const text = value.toString(8).padStart(length - 1, "0").slice(0, length - 1);
  buffer.write(text, offset, length - 1, "ascii");
  buffer[offset + length - 1] = 0;
}

function paddingFor(size) {
  const remainder = size % 512;
  return remainder === 0 ? 0 : 512 - remainder;
}

function verifyArchiveEntryList(entries) {
  const violations = [];
  for (const entry of entries) {
    const relPath = entry.relPath;
    const parts = relPath.split("/");
    const name = basename(relPath);
    if (relPath.startsWith("/") || parts.includes("..")) violations.push(relPath);
    if (isEnvFile(name)) violations.push(relPath);
    if (isPayloadLogName(relPath)) violations.push(relPath);
    if (parts.some((part) => blockedDirReasons.has(part))) violations.push(relPath);
  }
  if (violations.length > 0) {
    throw new Error(`Backup verification failed; blocked paths were included: ${violations.slice(0, 10).join(", ")}`);
  }
}

function timestamp() {
  const now = new Date();
  const pad = (value) => String(value).padStart(2, "0");
  return `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}-${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`;
}

function formatBytes(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

File: scripts/create_local_backup.py
#!/usr/bin/env python3
"""Create a local source backup without runtime state or secrets."""

from __future__ import annotations

import datetime as dt
import os
import re
import subprocess
import tarfile
from dataclasses import dataclass
from pathlib import Path


MAX_SCAN_BYTES = 5 * 1024 * 1024


BLOCKED_DIRS = {
    ".cache",
    ".git",
    ".hg",
    ".mypy_cache",
    ".next",
    ".nuxt",
    ".parcel-cache",
    ".playwright-cli",
    ".pytest_cache",
    ".ruff_cache",
    ".svn",
    ".turbo",
    ".venv",
    "__pycache__",
    "backups",
    "build",
    "coverage",
    "dist",
    "env",
    "exports",
    "htmlcov",
    "local_backups",
    "node_modules",
    "out",
    "tmp",
    "venv",
    "workspaces",
}

BLOCKED_FILE_SUFFIXES = {
    ".log",
    ".pyc",
    ".pyo",
    ".sqlite",
    ".sqlite-journal",
    ".sqlite-shm",
    ".sqlite-wal",
}

BLOCKED_FILE_NAMES = {
    ".DS_Store",
}

RAW_PROMPT_OR_PAYLOAD_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in (
        r"provider[-_ ]?payload",
        r"provider[-_ ]?response",
        r"provider[-_ ]?request",
        r"provider[-_ ]?prompt",
        r"raw[-_ ]?provider",
        r"raw[-_ ]?prompt",
        r"raw[-_ ]?response",
        r"prompt[-_ ]?payload",
        r"prompt[-_ ]?log",
        r"payload[-_ ]?log",
        r"ai[-_ ]?payload",
        r"llm[-_ ]?payload",
    )
]

SECRET_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in (
        r"\bsk-(?:proj-)?[A-Za-z0-9_-]{20,}\b",
        r"\bDEEPSEEK_API_KEY\s*=\s*[\"']?[A-Za-z0-9_-]{20,}",
        r"\bAuthorization\s*:\s*Bearer\s+[A-Za-z0-9._-]{20,}",
        r"\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{20,}\b",
        r"\bgithub_pat_[A-Za-z0-9_]{30,}\b",
        r"\bAKIA[0-9A-Z]{16}\b",
        r"\bxox[baprs]-[A-Za-z0-9-]{20,}\b",
    )
]

SAFE_PLACEHOLDER_MARKERS = {
    "example",
    "fake",
    "local-only",
    "placeholder",
    "test",
    "your_",
}

SAFE_PLACEHOLDER_PREFIXES = (
    "sk-deepseek-",
    "sk-live-",
    "sk-openai-",
    "sk-prototype-",
    "sk-release-",
    "sk-route-",
)


@dataclass
class OmissionCounts:
    env_files: int = 0
    git_internals: int = 0
    runtime_exports: int = 0
    local_caches: int = 0
    local_state: int = 0
    backup_recursion: int = 0
    raw_prompts_or_payloads: int = 0
    secret_matches: int = 0
    local_noise: int = 0
    symlinks: int = 0


def main() -> int:
    repo_root = find_repo_root()
    backup_dir = repo_root / "local_backups"
    backup_dir.mkdir(parents=True, exist_ok=True)

    archive_name = f"sprintos-local-backup-{timestamp()}.tar.gz"
    archive_path = backup_dir / archive_name
    counts = OmissionCounts()
    included_files = 0

    with tarfile.open(archive_path, "w:gz") as archive:
        for path in sorted(repo_root.rglob("*")):
            rel_path = path.relative_to(repo_root)
            rel_posix = rel_path.as_posix()
            reason = exclusion_reason(path, rel_path)
            if reason:
                increment(counts, reason)
                if path.is_dir():
                    continue
                continue
            if path.is_symlink():
                counts.symlinks += 1
                continue
            if path.is_dir():
                continue
            if path.is_file() and file_contains_secret(path):
                counts.secret_matches += 1
                continue
            add_file(archive, path, rel_posix)
            included_files += 1

    verify_archive(archive_path)
    archive_size = archive_path.stat().st_size

    print(f"Backup created: {archive_path.relative_to(repo_root)}")
    print(f"Archive size: {format_bytes(archive_size)}")
    print(f"Included files: {included_files}")
    print("Omitted files/folders:")
    print(f"- env files: {counts.env_files}")
    print(f"- git internals: {counts.git_internals}")
    print(f"- runtime exports/workspaces: {counts.runtime_exports}")
    print(f"- local caches/dependencies: {counts.local_caches}")
    print(f"- local SQLite/runtime state: {counts.local_state}")
    print(f"- backup recursion: {counts.backup_recursion}")
    print(f"- raw prompts/provider payloads: {counts.raw_prompts_or_payloads}")
    print(f"- files with obvious secret patterns: {counts.secret_matches}")
    print(f"- local noise/symlinks: {counts.local_noise + counts.symlinks}")
    print("Verification passed: archive entries contain no blocked env files, git internals, caches, runtime exports, raw prompts, provider payloads, local backups, or obvious secret paths.")
    return 0


def find_repo_root() -> Path:
    try:
        output = subprocess.check_output(
            ["git", "rev-parse", "--show-toplevel"],
            text=True,
            stderr=subprocess.DEVNULL,
        ).strip()
        if output:
            return Path(output).resolve()
    except (subprocess.CalledProcessError, FileNotFoundError):
        pass
    return Path.cwd().resolve()


def timestamp() -> str:
    return dt.datetime.now().strftime("%Y%m%d-%H%M%S")


def exclusion_reason(path: Path, rel_path: Path) -> str | None:
    parts = rel_path.parts
    name = path.name
    rel_posix = rel_path.as_posix()

    if is_env_file(name):
        return "env_files"
    if ".git" in parts:
        return "git_internals"
    if any(part in {"exports", "workspaces"} for part in parts):
        return "runtime_exports"
    if any(part in {"local_backups", "backups"} for part in parts):
        return "backup_recursion"
    if any(part in BLOCKED_DIRS for part in parts):
        return "local_caches"
    if parts and parts[0] == "data" and path.is_file():
        return "local_state"
    if any(pattern.search(rel_posix) for pattern in RAW_PROMPT_OR_PAYLOAD_PATTERNS):
        return "raw_prompts_or_payloads"
    if name in BLOCKED_FILE_NAMES or any(name.endswith(suffix) for suffix in BLOCKED_FILE_SUFFIXES):
        return "local_noise"
    return None


def is_env_file(name: str) -> bool:
    return name == ".env" or name.startswith(".env.")


def file_contains_secret(path: Path) -> bool:
    try:
        if path.stat().st_size > MAX_SCAN_BYTES:
            return False
        data = path.read_bytes()
    except OSError:
        return False
    if b"\x00" in data[:8000]:
        return False
    try:
        text = data.decode("utf-8")
    except UnicodeDecodeError:
        return False
    for pattern in SECRET_PATTERNS:
        for match in pattern.finditer(text):
            if not is_safe_placeholder_match(match.group(0)):
                return True
    return False


def is_safe_placeholder_match(value: str) -> bool:
    normalized = value.lower()
    if any(marker in normalized for marker in SAFE_PLACEHOLDER_MARKERS):
        return True
    token_match = re.search(r"sk-[a-z0-9_-]+", normalized)
    if token_match and token_match.group(0).startswith(SAFE_PLACEHOLDER_PREFIXES):
        return True
    return False


def add_file(archive: tarfile.TarFile, path: Path, rel_posix: str) -> None:
    info = archive.gettarinfo(str(path), arcname=rel_posix)
    info.uid = 0
    info.gid = 0
    info.uname = "sprintos"
    info.gname = "sprintos"
    with path.open("rb") as handle:
        archive.addfile(info, handle)


def verify_archive(archive_path: Path) -> None:
    violations: list[str] = []
    with tarfile.open(archive_path, "r:gz") as archive:
        for member in archive.getmembers():
            member_path = Path(member.name)
            parts = member_path.parts
            name = member_path.name
            if member.name.startswith("/") or ".." in parts:
                violations.append(member.name)
            if is_env_file(name):
                violations.append(member.name)
            if ".git" in parts:
                violations.append(member.name)
            if any(part in {"exports", "workspaces", "local_backups", "backups"} for part in parts):
                violations.append(member.name)
            if any(part in BLOCKED_DIRS for part in parts):
                violations.append(member.name)
            if any(pattern.search(member.name) for pattern in RAW_PROMPT_OR_PAYLOAD_PATTERNS):
                violations.append(member.name)
    if violations:
        sample = ", ".join(violations[:10])
        raise RuntimeError(f"Backup verification failed; blocked entries were included: {sample}")


def increment(counts: OmissionCounts, field: str) -> None:
    setattr(counts, field, getattr(counts, field) + 1)


def format_bytes(size: int) -> str:
    if size < 1024:
        return f"{size} B"
    if size < 1024 * 1024:
        return f"{size / 1024:.1f} KB"
    return f"{size / (1024 * 1024):.1f} MB"


if __name__ == "__main__":
    raise SystemExit(main())
